File: LeetCodeTasks/25_reverse-nodes-in-k-group/solutions/cpp/solution.cpp
#include <vector>

#include <algorithm>


class Solution {

public:

    ListNode* reverseKGroup(ListNode* head, int k) {

        if (!head || k == 1) return head;



        ListNode dummy(0);

        dummy.next = head;

        ListNode *prev = &dummy, *curr = head, *nex = nullptr;



        int len = 0;

        while (curr) { // Calculate length of list

            len++;

            curr = curr->next;

        }



        while (len >= k) {

            curr = prev->next;

            nex = curr->next;

            for (int i = 1; i < k; ++i) {

                curr->next = nex->next;

                nex->next = prev->next;

                prev->next = nex;

                nex = curr->next;

            }

            prev = curr;

            len -= k;

        }



        return dummy.next;

    }

};

File: LeetCodeTasks/25_reverse-nodes-in-k-group/solutions/python3/solution.py

class Solution:

    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:

        if head is None or k == 1:

            return head

        

        # Dummy node to handle edge cases easily

        dummy = ListNode(-1)

        dummy.next = head

        prev_end = dummy

        current = head

        

        # Helper function to reverse k nodes

        def reverse(start, k):

            prev, node = None, start

            for _ in range(k):

                next_node = node.next

                node.next = prev

                prev = node

                node = next_node

            return prev, start  # New head and tail after reversal

        

        # Count nodes

        count = 0

        while current:

            count += 1

            if count % k == 0:

                next_group = current.next

                # Reverse the k nodes

                new_head, new_tail = reverse(prev_end.next, k)

                prev_end.next = new_head

                new_tail.next = next_group

                prev_end = new_tail

                current = next_group

            else:

                current = current.next



        return dummy.next

